imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -0,0 +1,259 @@
+        ----------
+        seconds : int
+            Number of seconds to add
+
+        Returns
+        -------
+        time : TimeTuple
+            New time tuple with the current time tuple + seconds.
+        """
+        # Add whole seconds to coarse time
+        coarse = self.coarse_time + floor(seconds)
+        # fine time is 1/65535th of a second
+        fine = self.fine_time + round((seconds % 1) * MAX_FINE_TIME)
+
+        # If fine is larger than the max, move the excess into coarse time.
+        if fine > MAX_FINE_TIME:
+            coarse = coarse + floor(fine / MAX_FINE_TIME)
+            fine = fine % MAX_FINE_TIME
+
+        return TimeTuple(coarse, fine)
+
+
+@dataclass
+class Vector:
+    """
+    Data class for storing MAG vector data.
+
+    Attributes
+    ----------
+    timestamp : TimeTuple
+        Time of the vector sample
+    vectors : tuple[int, int, int, int]
+        Vector sample, containing x,y,z,range
+    """
+
+    # TODO: This timestamp should be in J2000/datetime64
+    timestamp: TimeTuple
+    x: int
+    y: int
+    z: int
+    rng: int
+
+    def __init__(self, vectors, time: TimeTuple):
+        self.timestamp = time
+        self.x = vectors[0]
+        self.y = vectors[1]
+        self.z = vectors[2]
+        self.rng = vectors[3]
+
+
+@dataclass
+class MagL1a:
+    """
+    Data class for MAG Level 1A data.
+
+    Attributes
+    ----------
+    is_mago : bool
+        True if the data is from MagO, False if data is from MagI
+    active : bool
+        True if the sensor is active
+    start_time : TimeTuple
+        The coarse and fine time for the sensor
+    vectors_per_second : int
+        Number of vectors per second
+    expected_vector_count : int
+        Expected number of vectors (vectors_per_second * seconds_of_data)
+    seconds_of_data : int
+        Number of seconds of data
+    SHCOARSE : int
+        Mission elapsed time
+    vectors : list[Vector]
+        List of magnetic vector samples, starting at start_time
+
+    """
+
+    is_mago: bool
+    active: bool
+    start_time: TimeTuple
+    vectors_per_second: int
+    expected_vector_count: int
+    seconds_of_data: int
+    SHCOARSE: int
+    vectors: list
+
+    def __post_init__(self):
+        """
+        Convert the vector list to a vector list with timestamps associated.
+
+        The first vector starts at start_time, then each subsequent vector time is
+        computed by adding 1/vectors_per_second to the previous vector's time.
+
+        This replaces self.vectors with a list of Vector objects.
+        """
+        sample_time_interval = 1 / self.vectors_per_second
+        previous_time = self.start_time
+        for index, vector in enumerate(self.vectors):
+            if index == 0:
+                new_vector = Vector(vector, self.start_time)
+            else:
+                new_vector = Vector(vector, previous_time + sample_time_interval)
+
+            previous_time = new_vector.timestamp
```

Comment:
I think this can be simplified a bit by updating the time at the end of your loop iteration.
```suggestion
        sample_time_interval = 1 / self.vectors_per_second
        current_time = self.start_time
        for index, vector in enumerate(self.vectors):
            new_vector = Vector(vector, current_time)
            # Increment our time counter before the next iteration
            current_time = new_vector.timestamp + sample_time_interval
```

Reply:
Nice, that's way better